test/server/cards/01_SOR/events/HeroicSacrifice.spec.ts

```diff
@@ -0,0 +1,60 @@
+describe('Heroic Sacrifice', function() {
+    integration(function(contextRef) {
+        describe('Heroic Sacrifice\'s ability', function() {
+            beforeEach(function () {
+                contextRef.setupTest({
+                    phase: 'action',
+                    player1: {
+                        hand: ['heroic-sacrifice'],
+                        groundArena: ['isb-agent', { card: 'wampa', exhausted: true }],
+                        spaceArena: ['tieln-fighter']
+                    },
+                    player2: {
+                        groundArena: ['sundari-peacekeeper', 'atst', { card: 'crafty-smuggler', upgrades: ['shield'] }],
+                    }
+                });
+            });
+
+            it('should draw a card and attack with a unit giving +2/+0 for this attack and defeat it after dealing combat damage', function () {
+                const { context } = contextRef;
+
+                context.player1.clickCard(context.heroicSacrifice);
+
+                expect(context.player1.handSize).toBe(1);
+
+                expect(context.player1).toBeAbleToSelectExactly([context.isbAgent, context.tielnFighter]);
+
+                context.player1.clickCard(context.isbAgent);
+                expect(context.player1).toBeAbleToSelectExactly([context.sundariPeacekeeper, context.atst, context.craftySmuggler, context.p2Base]);
+
+                context.player1.clickCard(context.p2Base);
+                expect(context.p2Base.damage).toBe(3);
+                expect(context.isbAgent).toBeInZone('discard');
+
+                expect(context.player2).toBeActivePlayer();
+            });
+
+            it('should draw a card and attack with a unit giving +2/+0 for this attack and not defeat unit if no combat damage is dealt', function () {
+                const { context } = contextRef;
+
+                context.player1.clickCard(context.heroicSacrifice);
+
+                expect(context.player1.handSize).toBe(1);
+
+                expect(context.player1).toBeAbleToSelectExactly([context.isbAgent, context.tielnFighter]);
+
+                context.player1.clickCard(context.isbAgent);
+                expect(context.player1).toBeAbleToSelectExactly([context.sundariPeacekeeper, context.atst, context.craftySmuggler, context.p2Base]);
+
+                context.player1.clickCard(context.craftySmuggler);
+
+                expect(context.craftySmuggler.upgrades.length).toBe(0);
+                expect(context.craftySmuggler).toBeInZone('groundArena');
+
+                expect(context.isbAgent).toBeInZone('groundArena');
+
+                expect(context.player2).toBeActivePlayer();
+            });
```

Comment:
Here's a fun test we can do to confirm that it is specifically combat damage that works and not damage from e.g. on-attack abilities:

1. Set up Lom Pyke with a Vambrace Flamethrower
2. Have Lom attack into a unit. You'll get a prompt to select which ability to trigger (see TriggeredAbilityWindow.spec.ts for an example of how to manage that)
3. First select Vambrace Flamethrower and deal damage to the defender and other units
4. Then activate Lom Pyke's ability and give a shield to the defender
5. Lom's combat damage won't resolve b/c of the shield but he should stay in play and not be defeated by Heroic Sacrifice b/c the Flamethrower damage wasn't combat damage

Reply:
Love the suggestion! That's a great scenario